feat: blog about html-eslint

yeonjuan opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [JS Foundation Code of Conduct](https://eslint.org/conduct).
-->

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request?
<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/latest/contribute/pull-requests)
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)

Add blog post about html-eslint

#### Related Issues

#### Is there anything you'd like reviewers to focus on?

<!-- markdownlint-disable-file MD004 -->

nzakas commented on src/_includes/partials/author_bios/yeonjuan.md:
```diff
@@ -0,0 +1 @@
+@eslint team alumnus. Passionate about JavaScript-based developer tools.
```
Just to double-check, would you like to expand your bio here? This is pulled from GitHub but you can definitely add more information if you'd like.

nzakas commented on src/content/drafts/eslint-html-plugin.md:
```diff
@@ -0,0 +1,119 @@
+---
+layout: post
+title: html-eslint now supports ESLint language plugin
```
How about this to make it a bit more ESLint-centric?
```suggestion
title: ESLint can now lint HTML using the html-eslint language plugin
```

nzakas commented on src/content/drafts/eslint-html-plugin.md:
```diff
@@ -0,0 +1,119 @@
+---
+
+In 2024, [ESLint announced](/blog/2024/07/whats-coming-next-for-eslint/) its plan to become a language-agnostic linter that is capable of linting languages other than JavaScript.
+Since then, we've seen official support added for JSON, Markdown, and most recently, CSS.
```
```suggestion
Since then, we've seen official support added for [JSON and Markdown](https://eslint.org/blog/2024/10/eslint-json-markdown-support/), and most recently, [CSS](https://eslint.org/blog/2025/02/eslint-css-support/).
```

nzakas commented on src/content/drafts/eslint-html-plugin.md:
```diff
@@ -0,0 +1,119 @@
+
+In 2024, [ESLint announced](/blog/2024/07/whats-coming-next-for-eslint/) its plan to become a language-agnostic linter that is capable of linting languages other than JavaScript.
+Since then, we've seen official support added for JSON, Markdown, and most recently, CSS.
+Today, I'm excited to share that [html-eslint](https://html-eslint.org) now supports ESLint's language plugin for HTML.
```
```suggestion
Today, I'm excited to share that [html-eslint](https://html-eslint.org) now provides an ESLint language plugin for linting HTML.
```

nzakas commented on src/content/drafts/eslint-html-plugin.md:
```diff
@@ -0,0 +1,119 @@
+Since then, we've seen official support added for JSON, Markdown, and most recently, CSS.
+Today, I'm excited to share that [html-eslint](https://html-eslint.org) now supports ESLint's language plugin for HTML.
+
+## HTML linting with `html-eslint`
```
```suggestion
## Linting HTML with `html-eslint`
```

nzakas commented on src/content/drafts/eslint-html-plugin.md:
```diff
@@ -0,0 +1,119 @@
+
+## HTML linting with `html-eslint`
+
+The [`@html-eslint/eslint-plugin`](https://www.npmjs.com/package/@html-eslint/eslint-plugin) plugin provides support for linting HTML code inside `.html` files. You can now configure HTML linting using ESLint's flat config and language plugin system. To get started, install it from npm:
```
```suggestion
The [`@html-eslint/eslint-plugin`](https://www.npmjs.com/package/@html-eslint/eslint-plugin) package provides support for linting HTML code inside `.html` files. To get started, install it from npm:
```

nzakas commented on src/content/drafts/eslint-html-plugin.md:
````diff
@@ -0,0 +1,119 @@
+Then update your configuration file:
+
+```js
+import html from "@html-eslint/eslint-plugin";
+
+export default [
+    // lint html files
+    {
+        files: ["**/*.html"],
+        plugins: {
+            html,
+        },
+        language: "html/html",
+        rules: {
+            "html/no-duplicate-class": "error",
+        }
+    }
+]
````
Let's be sure to use the new `defineConfig()` helper here.
```suggestion
import { defineConfig } from "eslint/config";
import html from "@html-eslint/eslint-plugin";

export default defineConfig([
    // lint html files
    {
        files: ["**/*.html"],
        plugins: {
            html,
        },
        language: "html/html",
        rules: {
            "html/no-duplicate-class": "error",
        }
    }
]);
```

nzakas commented on src/content/drafts/eslint-html-plugin.md:
````diff
@@ -0,0 +1,119 @@
+]
+```
+
+The plugin includes a growing list of rules covering best practices, code styles, SEO and accessibility in HTML. You can find the list of supported rules [here](https://html-eslint.org/docs/rules).
+
+Have an idea for a rule? [Open an issue](https://github.com/yeonjuan/html-eslint/issues).
````
```suggestion
The plugin includes a growing list of [rules](https://html-eslint.org/docs/rules)) covering best practices, code style, search engine optimization (SEO), and accessibility. (Have an idea for a rule? [Open an issue](https://github.com/yeonjuan/html-eslint/issues).)
```

nzakas commented on src/content/drafts/eslint-html-plugin.md:
```diff
@@ -0,0 +1,119 @@
+
+## Template engine syntax support
+
+Rather than using plain HTML, some projects use template engines like Handlebars, which introduce non-standard syntax such as {% raw %}`{{variable}}`{% endraw %}.
```
```suggestion
Rather than using plain HTML, some projects use template engines like [Handlebars](https://handlebarsjs.com/), which introduce non-standard syntax such as {% raw %}`{{variable}}`{% endraw %}.
```

nzakas commented on src/content/drafts/eslint-html-plugin.md:
```diff
@@ -0,0 +1,119 @@
+
+Rather than using plain HTML, some projects use template engines like Handlebars, which introduce non-standard syntax such as {% raw %}`{{variable}}`{% endraw %}.
+
+`html-eslint` is designed to offer support for custom syntax patterns used in popular template engines. This will allow users to configure html-eslint to tolerate these syntaxes.
```
```suggestion
`html-eslint` is designed to offer support for custom syntax patterns used in popular template engines. This allows users to configure html-eslint to tolerate these syntaxes. Here's an example:
```

nzakas commented on src/content/drafts/eslint-html-plugin.md:
````diff
@@ -0,0 +1,119 @@
+
+`html-eslint` is designed to offer support for custom syntax patterns used in popular template engines. This will allow users to configure html-eslint to tolerate these syntaxes.
+
+```js
+import html from "@html-eslint/eslint-plugin";
+
+export default [
+    // lint html files
+    {
+        files: ["**/*.html"],
+        plugins: {
+            html,
+        },
+        language: "html/html",
+        languageOptions: {
+            // This tells the parser to treat {% raw %}{{ ... }}{% endraw %} as template syntax,
+            // so it won’t try to parse contents inside as regular HTML
+            templateEngineSyntax: {
+                {% raw %}"{{": "}}"{% endraw %},
+            },
+        },
+        rules: {
+            "html/require-img-alt": "error"
+        }
+    }
+]
+```
````
```suggestion
```js
import { defineConfig } from "eslint/config";
import html from "@html-eslint/eslint-plugin";

export default defaultConfig([
    // lint html files
    {
        files: ["**/*.html"],
        plugins: {
            html,
        },
        language: "html/html",
        languageOptions: {
            // This tells the parser to treat {% raw %}{{ ... }}{% endraw %} as template syntax,
            // so it won’t try to parse contents inside as regular HTML
            templateEngineSyntax: {
                {% raw %}"{{": "}}"{% endraw %},
            },
        },
        rules: {
            "html/require-img-alt": "error"
        }
    }
]);
```
```

nzakas commented on src/content/drafts/eslint-html-plugin.md:
````diff
@@ -0,0 +1,119 @@
+]
+```
+
+## Linting HTML code inside JavaScript Template Literal
````
```suggestion
## Linting HTML code inside JavaScript template literals
```

nzakas commented on src/content/drafts/eslint-html-plugin.md:
````diff
@@ -0,0 +1,119 @@
+To enable this, you don’t need to set a language. Just apply `html-eslint` rules to your JavaScript or TypeScript files, and the plugin will detect and lint HTML within template literals.
+
+```js
+import html from "@html-eslint/eslint-plugin";
+
+export default [
+    {
+        files: ["**/*.js", "**/*.ts"],
+        plugins: {
+            html,
+        },
+        rules: {
+            "html/require-img-alt": "error",
+        },
+    },
+];
````
```suggestion
import { defineConfig } from "eslint/config";
import html from "@html-eslint/eslint-plugin";

export default defineConfig([
    {
        files: ["**/*.js", "**/*.ts"],
        plugins: {
            html,
        },
        rules: {
            "html/require-img-alt": "error",
        },
    },
]);
```

nzakas commented on src/content/drafts/eslint-html-plugin.md:
```diff
@@ -0,0 +1,119 @@
+
+This is especially useful when working with libraries like [Lit](https://lit.dev/), where HTML is commonly written inside tagged template literals.
+
+To enable this, you don’t need to set a language. Just apply `html-eslint` rules to your JavaScript or TypeScript files, and the plugin will detect and lint HTML within template literals.
```
This is really cool!

nzakas reviewed (commented):
Really nice work on this. I love all the functionality of html-eslint. :heart:

I left suggestions throughout, mostly for cleaning up language, providing more links to references, and updating the code examples to use `defineConfig()`.

A couple high-level things:

1. You use both html-eslint and `html-eslint` as the project name. Please update to use one or the other for consistency in the post.
2. It would be good to add HTML parsing to [Code Explorer](https://github.com/eslint/code-explorer) so that people can see the AST that it parses into. Do you want to submit a PR to do that?

Thanks again, this is really awesome work.

yeonjuan commented:
@nzakas Thank you! I've addressed the review comments

> You use both html-eslint and `html-eslint` as the project name. Please update to use one or the other for consistency in the post.

I've updated it to consistently use `html-eslint`.

It would be good to add HTML parsing to [Code Explorer](https://github.com/eslint/code-explorer) so that people can see the AST that it parses into. Do you want to submit a PR to do that?

Certainly!. Do you think this should be completed before the blog post is published?

nzakas commented on src/_includes/partials/author_bios/yeonjuan.md:
```diff
@@ -1 +1 @@
-@eslint team alumnus. Passionate about JavaScript-based developer tools.
\ No newline at end of file
+Yeon Juan is a software developer at [LINE Plus](https://www.linepluscorp.com/) and a @eslint team alumnus. I'm passionate about building JavaScript-based developer tools that improve code quality and the developer experience.
```
Just to keep it in the third person.
```suggestion
Yeon Juan is a software developer at [LINE Plus](https://www.linepluscorp.com/) and an ESLint team alumnus. He is passionate about building JavaScript-based developer tools that improve code quality and the developer experience.
```

nzakas commented:
> Certainly!. Do you think this should be completed before the blog post is published?

Yes, I think it would be nice to include a link in the post to an HTML example in Code Explorer.

lumirlumir commented on src/content/drafts/eslint-html-plugin.md:
````diff
@@ -0,0 +1,119 @@
+]
+```
+
+The plugin includes a growing list of [rules](https://html-eslint.org/docs/rules)) covering best practices, code style, search engine optimization (SEO), and accessibility. (Have an idea for a rule? [Open an issue](https://github.com/yeonjuan/html-eslint/issues).)
````
```suggestion
The plugin includes a growing list of [rules](https://html-eslint.org/docs/rules) covering best practices, code style, search engine optimization (SEO), and accessibility. (Have an idea for a rule? [Open an issue](https://github.com/yeonjuan/html-eslint/issues).
```

I think the `)` might be a mistake.

lumirlumir commented on src/content/drafts/eslint-html-plugin.md:
```diff
@@ -0,0 +1,119 @@
+
+In 2024, [ESLint announced](/blog/2024/07/whats-coming-next-for-eslint/) its plan to become a language-agnostic linter that is capable of linting languages other than JavaScript.
+Since then, we've seen official support added for [JSON and Markdown](https://eslint.org/blog/2024/10/eslint-json-markdown-support/), and most recently, [CSS](https://eslint.org/blog/2025/02/eslint-css-support/).
+Today, I'm excited to share that [html-eslint](https://html-eslint.org) now provides an ESLint language plugin for linting HTML.
```
```suggestion
Today, I'm excited to share that [`html-eslint`](https://html-eslint.org) now provides an ESLint language plugin for linting HTML.
```

Would it be better to wrap `html-eslint` with backticks for consistency?

lumirlumir reviewed (changes requested):
It looks really good to me! 

Just a few minor typos, I’ve left some comments.

yeonjuan commented:
Hi @nzakas @lumirlumir 
I’ve created a PR for code-explorer. I look forward to your review!
https://github.com/eslint/code-explorer/pull/99

nzakas reviewed (approved):
This looks great, thanks! I'm going to get this published today!
